### PR TITLE
Refactor mail address factory to non-static approach

### PR DESCRIPTION
**Pull request by legionth**
Currently the `getByPrefix` is static method. 

This PR replace all the Code with an OO-designed approach and makes it easier testable.

### CODE REVIEW

File: Services/Mail/test/GroupNameAsMailValidatorTest.php [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+class GroupNameAsMailValidatorTest extends \ilMailBaseTest
```
Comment (mjansenDatabay) [inline Services/Mail/test/GroupNameAsMailValidatorTest.php]: Please add the "il" class prefix.

File: Services/Mail/test/GroupNameAsMailValidatorTest.php [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<?php
+
```
Comment (mjansenDatabay) [inline Services/Mail/test/GroupNameAsMailValidatorTest.php]: Please add the license information.

File: Services/Mail/classes/Address/Validator/GroupNameAsMailValidator.php [hunk 3]
```diff
@@ -0,0 +1,41 @@
+<?php
+
+class GroupNameAsMailValidator
```
Comment (mjansenDatabay) [inline Services/Mail/classes/Address/Validator/GroupNameAsMailValidator.php]: Please add the "il" class prefix.

File: Services/Mail/classes/Address/Validator/GroupNameAsMailValidator.php [hunk 4]
```diff
@@ -0,0 +1,41 @@
+<?php
+
```
Comment (mjansenDatabay) [inline Services/Mail/classes/Address/Validator/GroupNameAsMailValidator.php]: Please add the license information.

File: Services/Mail/classes/Address/Type/class.ilMailAddressTypeFactory.php [hunk 5]
```diff
@@ -7,39 +7,39 @@
  */
 class ilMailAddressTypeFactory
 {
+	private $groupNameValidator;
```
Comment (mjansenDatabay) [inline Services/Mail/classes/Address/Type/class.ilMailAddressTypeFactory.php]: Please add PHPdoc comments.

File: Services/Mail/classes/Address/Type/class.ilMailAddressTypeFactory.php [hunk 6]
```diff
@@ -7,39 +7,39 @@
...
+
+	public function __construct(GroupNameAsMailValidator $groupNameValidator = null)
```
Comment (mjansenDatabay) [inline Services/Mail/classes/Address/Type/class.ilMailAddressTypeFactory.php]: Please add PHPdoc comments.

File: Services/Mail/classes/class.ilMail.php [hunk 7]
```diff
@@ -111,16 +111,23 @@ class ilMail
 	/** @var ilObjUser[] */
 	protected static $userInstances = array();
 
+	private $mailAddressTypeFactory;
```
Comment (mjansenDatabay) [inline Services/Mail/classes/class.ilMail.php]: Please add PHPdoc comments.

File: Services/Mail/classes/class.ilMail.php [hunk 8]
```diff
@@ -111,16 +111,23 @@ class ilMail
...
+
 	/**
 	 * @param integer $a_user_id
 	 */
-	public function __construct($a_user_id)
+	public function __construct($a_user_id, ilMailAddressTypeFactory $mailAddressTypeFactory = null)
```
Comment (mjansenDatabay) [inline Services/Mail/classes/class.ilMail.php]: Please change the PHPdoc comments.

File: Services/Mail/test/ilServicesMailSuite.php [hunk 9]
```diff
@@ -28,6 +28,9 @@ public static function suite()
 		require_once 'Services/Mail/test/ilMailTransportSettingsTest.php';
 		$suite->addTestSuite('ilMailTransportSettingsTest');
 
+		require_once 'Services/Mail/test/GroupNameAsMailValidatorTest.php';
```
Comment (mjansenDatabay) [inline Services/Mail/test/ilServicesMailSuite.php]: The "il" prefix is still missing for the filename.

File: Services/Mail/classes/Address/Type/class.ilMailAddressTypeFactory.php [hunk 10]
```diff
@@ -7,39 +7,43 @@
  */
 class ilMailAddressTypeFactory
 {
+	/** @var GroupNameAsMailValidator */
+	private $groupNameValidator;
+
+	/**
+	 * @param GroupNameAsMailValidator|null $groupNameValidator
+	 */
+	public function __construct(GroupNameAsMailValidator $groupNameValidator = null)
```
Comment (mjansenDatabay) [inline Services/Mail/classes/Address/Type/class.ilMailAddressTypeFactory.php]: The "il" prefix is still missing for the type hint.

File: Services/Mail/classes/Address/Type/class.ilMailAddressTypeFactory.php [hunk 11]
```diff
@@ -7,39 +7,43 @@
  */
 class ilMailAddressTypeFactory
 {
+	/** @var GroupNameAsMailValidator */
```
Comment (mjansenDatabay) [inline Services/Mail/classes/Address/Type/class.ilMailAddressTypeFactory.php]: The "il" prefix is still missing here.

File: Services/Mail/classes/Address/Type/class.ilMailAddressTypeFactory.php [hunk 12]
```diff
@@ -7,39 +7,43 @@
...
+	private $groupNameValidator;
+
+	/**
+	 * @param GroupNameAsMailValidator|null $groupNameValidator
+	 */
+	public function __construct(GroupNameAsMailValidator $groupNameValidator = null)
+	{
+		if ($groupNameValidator === null) {
+			$groupNameValidator = new GroupNameAsMailValidator(ilMail::ILIAS_HOST);
```
Comment (mjansenDatabay) [inline Services/Mail/classes/Address/Type/class.ilMailAddressTypeFactory.php]: Please add the "il" prefix here...

File: Services/Mail/test/ilGroupNameAsMailValidatorTest.php [hunk 13]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+class ilGroupNameAsMailValidatorTest extends \ilMailBaseTest
+{
+	public function testObjectCanBeCreated()
+	{
+		$validator = new GroupNameAsMailValidator('someHost');
```
Comment (mjansenDatabay) [inline Services/Mail/test/ilGroupNameAsMailValidatorTest.php]: The "il" prefix is missing here.

**Review by mjansenDatabay — CHANGES_REQUESTED** [review]

File: Services/Mail/classes/Address/Type/class.ilMailRoleAddressType.php [hunk 14]
```diff
@@ -179,18 +180,23 @@ public function resolve()
...
+	 * @internal param IETF $string RFX 822 address list containing role mailboxes.
 	 */
-	public static function searchRolesByMailboxAddressList($a_address_list)
+	public static function searchRolesByMailboxAddressList($a_address_list, ilMailRfc822AddressParserFactory $parser = null)
 	{
 		global $DIC;
 
-		$role_ids = array();
+		if ($parser === null) {
+			$parser = ilMailRfc822AddressParserFactory::getParser($a_address_list);
```
Comment (mjansenDatabay) [inline Services/Mail/classes/Address/Type/class.ilMailRoleAddressType.php:194]: One last thing. Maybe we should switch to an instance based approach of for this factory as well.

**Review by mjansenDatabay — CHANGES_REQUESTED** [review]

File: Services/Mail/classes/Address/Type/class.ilMailRoleAddressType.php (hunk 14 above)
Comment (legionth) [reply Services/Mail/classes/Address/Type/class.ilMailRoleAddressType.php:194]: Good point! I will open a seperated PR for this :+1:

**Pull request merged**